src/V3Assert.cpp

```diff
@@ -353,10 +360,39 @@ class AssertVisitor final : public VNVisitor {
         }
         nodep->replaceWith(inp);
     }
+
+    //========== Move $sampled down to read-only variables
     virtual void visit(AstSampled* nodep) override {
+        if (nodep->user1()) return;
+        VL_RESTORER(m_inSampled);
+        {
+            m_inSampled = true;
+            iterateChildren(nodep);
+        }
         nodep->replaceWith(nodep->exprp()->unlinkFrBack());
         VL_DO_DANGLING(pushDeletep(nodep), nodep);
     }
+    virtual void visit(AstVarRef* nodep) override {
+        iterateChildren(nodep);
+        if (!m_inSampled) return;
```

Comment:
Suggest wrap below in if (m_inSampled) rather then short circuiting code, as at some point we may need to do other things on an AstVarRef and it'll be easier to avoid changing the code.

Reply:
Done (23aa37d602967452065b4c2f819a47cedb1a655f).